### ISSUE TITLE
ELMo padding using vocab_to_cache

### ISSUE THREAD

**Issue by coylz** [allenai/allennlp#2216]
Hello,

First, I've already seen issue #1810  but it does not answer my question.

I am using ELMo as a replacement for W2V, GloVe, etc. in my PyTorch model. 

For performance and convenience purpose, and because my input vocabulary is fixed during training, I do not have clear text in my training data but indices and a vocabulary as a list of strings.

With the classic torch.nn.Embedding class of PyTorch, I use the parameter "padding_idx=0", so I know that in my vocabulary, the word at index 0 (which is a special token "\<pad\>") will always have a fixed representation full of zeros. 

With the class ELMo, I use the "vocab_to_cache" feature, and the padding mechanism is unclear to me. 

Looking at the source code: the constructor of "ELMo" calls the constructor of "_ElmoBiLm", which calls the method "create_cached_cnn_embeddings", which calls the "Embedding" class with the parameter "padding_index=0".

I don't understand all the code so I might be wrong, but it seems that in the end my word at index 0 will be used as a special padding token, and this is exactly what I would like. But it is stated nowhere in the documentation so I'm really not sure.

Can you confirm that when we cache a vocabulary, the word at index 0 will be used for padding ? If not, can you explain how to do padding with vocab_to_cache ?

Thank you a lot.

**Comment by matt-peters** [comment]
We use 0 as the padding id: (https://github.com/allenai/allennlp/blob/master/allennlp/modules/elmo.py#L683)

Our embedding code also sets these embeddings to 0, although I wouldn't rely on it as this is an implementation detail and subject to change.  If your downstream code assumes the embeddings for PAD are 0.0, then I'd recommend explicitly zeroing it out yourself.
